src/main/java/com/liferay/faces/util/cache/CacheFactory.java

```diff
@@ -0,0 +1,228 @@
+	 * @param  externalContext  The external context associated with the current faces context.
+	 */
+	public static <K, V> Cache<K, V> getConcurrentCacheInstance(ExternalContext externalContext) {
+
+		CacheFactory cacheFactory = (CacheFactory) FactoryExtensionFinder.getFactory(externalContext,
+				CacheFactory.class);
+
+		return cacheFactory.<K, V>getConcurrentCache();
+	}
+
+	/**
+	 * Returns a new instance of {@link Cache} from the {@link CacheFactory} found by the {@link
+	 * FactoryExtensionFinder}. The returned instance is designed to be accessed and modified by multiple threads
+	 * concurrently, so it is guaranteed to be {@link java.io.Serializable}.
+	 *
+	 * @param  <K>              The type of the cache's keys.
+	 * @param  <V>              The type of the cache's values.
+	 * @param  externalContext  The external context associated with the current faces context.
+	 * @param  initialCapacity  The initial capacity of the cache.
+	 */
+	public static <K, V> Cache<K, V> getConcurrentCacheInstance(ExternalContext externalContext, int initialCapacity) {
+
+		CacheFactory cacheFactory = (CacheFactory) FactoryExtensionFinder.getFactory(externalContext,
+				CacheFactory.class);
+
+		return cacheFactory.<K, V>getConcurrentCache(initialCapacity);
+	}
+
+	/**
+	 * Returns a new instance of {@link Cache} from the {@link CacheFactory} found by the {@link
+	 * FactoryExtensionFinder}. The returned instance is designed to be accessed and modified by multiple threads
+	 * concurrently, so it is guaranteed to be {@link java.io.Serializable}. The returned cache will avoid exceeding the
+	 * maximum cache capacity. The algorithm used to ensure that the cache size does not exceed the max capacity (for
+	 * example least-recently-used) may cause significant performance differences between the returned cache and a cache
+	 * returned from {@link #getConcurrentCache()} or {@link #getConcurrentCache(int)}.
+	 *
+	 * @param  <K>              The type of the cache's keys.
+	 * @param  <V>              The type of the cache's values.
+	 * @param  externalContext  The external context associated with the current faces context.
+	 * @param  initialCapacity  The initial capacity of the cache.
+	 * @param  maxCapacity      The maximum capacity of the cache.
+	 */
+	public static <K, V> Cache<K, V> getConcurrentCacheInstance(ExternalContext externalContext, int initialCapacity,
+		int maxCapacity) {
+
+		CacheFactory cacheFactory = (CacheFactory) FactoryExtensionFinder.getFactory(externalContext,
+				CacheFactory.class);
+
+		return cacheFactory.<K, V>getConcurrentCache(initialCapacity, maxCapacity);
+	}
+
+	/**
+	 * Returns a new instance of {@link Cache}. The returned instance is designed to be used during execution of a
+	 * request thread, so it is not guaranteed to be {@link java.io.Serializable}.
+	 *
+	 * @param  <K>  The type of the cache's keys.
+	 * @param  <V>  The type of the cache's values.
+	 */
+	public abstract <K, V> Cache<K, V> getCache();
```

Comment:
@stiemannkj1 The [JavaDoc for java.util.HashMap](https://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html) indicates that a HashMap implements Serializable. So I think it is OK to say that the return value is not thread-safe, but is indeed guaranteed to be Serializable. The warning about "not guaranteed" seems like it might be loosely tied to the notion of "since this is not thread-safe, don't assume it is Serializable, which means don't put it in the session since session objects need to be Serializable to replicate." But in the final analysis I think that the two items are separate/distinct: thread-safety, and serializability.